### DISCUSSION TITLE
Can the random seed be fixed to get reproducible training results?

### DISCUSSION THREAD

**Discussion by davidndrio** [brucefan1983/GPUMD#1526] [Q&A] [answered]
Hi,

I noticed that running the same qNEP training twice with identical nep.in and training data gives substantially different results each time: different loss curves, different final RMSE, different nep.txt. For example, the RMSE of energies changed by a factor of ~2 between two otherwise identical runs.

Is there a way to fix the random seed to make results reproducible across fresh runs?

**Comment by brucefan1983** [comment, marked as answer]
Could you show loss plots?

If you want to enforce reproduction, you can add `-DDEBUG` to `makefile`:

```
CFLAGS = -std=c++14 -O3 $(CUDA_ARCH) -DDEBUG
```

Then
```
make clean
make -j 
```

But due to the use of `atomicAdd()` function, two trainings will gradaully (perhaps only slightly) diverge with increasing training steps.

Also, this will cause MD part to be deterministic too, which is usually not what you want. So usually you need to remove `-DDEBUG` to compile `gpumd`.

**Reply by davidndrio** [reply]
Thank you for the information!

Here are the loss functions and RMSE of the 2 trainings.
<img width="2138" height="700" alt="train2_rmse" src="https://github.com/user-attachments/assets/cfe27b00-0abc-4dc4-a55e-7e18b513ff33" />
<img width="1895" height="647" alt="train1_rmse" src="https://github.com/user-attachments/assets/0abdb5a0-0b9c-4321-b4e3-8bb703b5ad7d" />
<img width="919" height="700" alt="train2_loss" src="https://github.com/user-attachments/assets/5797be59-5546-4550-ac8b-3c991fe3b555" />
<img width="907" height="706" alt="train1_loss" src="https://github.com/user-attachments/assets/90d091ed-a64d-4478-bd1a-354a9689d167" />

**Comment by brucefan1983** [comment]
Seem the second run got better accuracy of BEC (Z), but worse for potential. 

There was a bug for BEC training with non-fullbatch mode. Which GPUMD version did you use. 

Otherwise, I think your dataset have small range of energy and a RMSE of 1 meV/atom is actually not much different from 2 meV/atom in practice.

**Reply by davidndrio** [reply]
Both were in version 5.2.

Got it.

Thank you very much for the help!

**Comment by brucefan1983** [comment]
The bug was fixed in GPUMD-5.2, so that's not the reason. I think they are just due to randomess in the training process.
